src/os.rs

```diff
@@ -330,41 +289,89 @@ mod imp {
+    // is less random.
+    //
+    // As a countermeasure we try to do a single read from `/dev/random` in
+    // non-blocking mode. If the OS RNG is not yet properly seeded, we will get
+    // an error. Because we keep `/dev/urandom` open when succesful, this is
+    // only a small one-time cost.
+    fn open_dev_random() -> Result<(), Error> {
+        fn map_err(err: io::Error) -> Error {
+            match err.kind() {
+                io::ErrorKind::Interrupted =>
+                        Error::new(ErrorKind::Transient, "interrupted"),
+                io::ErrorKind::WouldBlock =>
+                        Error::with_cause(ErrorKind::NotReady,
+                        "OS RNG not yet seeded", err),
+                _ => Error::with_cause(ErrorKind::Unavailable,
+                        "error while opening random device", err)
+            }
+        }
 
-    #[derive(Debug)]
-    enum OsRngInner {
-        OsGetrandomRng,
-        OsReadRng(ReadRng),
-    }
+        READ_RNG_ONCE.call_once(|| {
+            unsafe { READ_RNG_FILE = Some(Mutex::new(None)) }
+        });
 
-    impl OsRng {
-        pub fn new() -> Result<OsRng, Error> {
-            if is_getrandom_available() {
-                return Ok(OsRng { inner: OsGetrandomRng });
+        // We try opening the file outside the `call_once` fn because we cannot
+        // clone the error, thus we must retry on failure.
+
+        let mutex = unsafe { READ_RNG_FILE.as_ref().unwrap() };
+        let mut guard = mutex.lock().unwrap();
+        if (*guard).is_none() {
+            {
+                info!("OsRng: opening random device /dev/random");
+                let mut file = OpenOptions::new()
+                    .read(true)
+                    .custom_flags(libc::O_NONBLOCK)
+                    .open("/dev/random")
+                    .map_err(|err| map_err(err))?;
+                let mut buf = [0u8; 1];
+                file.read_exact(&mut buf).map_err(|err| map_err(err))?;
             }
 
-            let reader_rng = ReadRng::open("/dev/urandom")?;
-            Ok(OsRng { inner: OsReadRng(reader_rng) })
-        }
-        
-        pub fn try_fill_bytes(&mut self, v: &mut [u8]) -> Result<(), Error> {
-            match self.inner {
-                OsGetrandomRng => getrandom_try_fill(v),
-                OsReadRng(ref mut rng) => rng.try_fill_bytes(v)
+            info!("OsRng: opening random device /dev/urandom");
+            let file = File::open("/dev/urandom").map_err(|err| map_err(err))?;
+            *guard = Some(file);
+        };
+        Ok(())
+    }
+
+    fn dev_random_try_fill(dest: &mut [u8]) -> Result<(), Error> {
+        if dest.len() == 0 { return Ok(()); }
+        trace!("OsRng: reading {} bytes from random device", dest.len());
+
+        // Since we have an instance of Self, we can assume that our memory was
+        // set with a valid object.
+        let mutex = unsafe{ READ_RNG_FILE.as_ref().unwrap() };
```

Comment:
This comment is incorrect; it was there in `ReadRng` because existence of the `self` parameter proves the constructor ran. Here we just have to use local-analysis to see that this function only gets called after `open_dev_random` succeeds.